### ISSUE TITLE
iOS VoiceOver accessibility regression: EditableText rotor actions only available when keyboard is focused

### ISSUE THREAD

**Issue by luomo-pro** [flutter/flutter#151029]
After the changes implemented in issue #94465, there seems to be a regression in VoiceOver accessibility for text editing on iOS.
While I cannot be certain that this issue directly caused the problem, I noticed the regression after this change was implemented, suggesting a potential correlation.
This issue significantly impacts the user experience for visually impaired users.

Current behavior:
1. When a user activates an editable text field and the keyboard appears, they must first focus on the keyboard with VoiceOver before using the rotor to access editing options like "Select All" and "Paste".
2. If VoiceOver focus is on the text field instead of the keyboard, using the rotor to navigate to "Edit" only shows "Scan Text" option.

Expected behavior (previous functionality):
Users could immediately use the VoiceOver rotor to access editing options like "Select All" and "Paste" after activating the text field, without needing to focus on the keyboard first.

This change makes it much more difficult for visually impaired users to perform common text editing tasks. Previously, they could easily paste text using the rotor immediately after activating the text field. Now, they must first navigate to the keyboard, which can be problematic, especially if the keyboard input mode is set to "Touch Typing". In this mode, any touch on the keyboard immediately activates keys, potentially inserting unwanted characters.

To reproduce:
1. Enable VoiceOver on an iOS device.
2. Run the following minimal example:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('VoiceOver Test')),
        body: const Center(
          child: Padding(
            padding: EdgeInsets.all(16.0),
            child: TextField(
              decoration: InputDecoration(
                hintText: 'Enter text here',
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

3. Activate the text field.
4. Attempt to use the VoiceOver rotor to access editing options.

Flutter version: Flutter (Channel master, 3.23.0-13.0.pre.298)

**Comment by darshankawar** [comment]
Thanks for the detailed report. I was able to replicate the reported behavior using master version only, and not in stable.

<details>
<summary> stable, master flutter doctor -v </summary>

```
[!] Flutter (Channel stable, 3.22.2, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.22.2 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (33 hours ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • Dart version 3.4.3
    • DevTools version 2.34.3
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.23.0-13.0.pre.357, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.23.0-13.0.pre.357 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 805d79607e (87 minutes ago), 2024-06-27 05:53:58 +0200
    • Engine revision a9194f0f01
    • Dart version 3.5.0 (build 3.5.0-304.0.dev)
    • DevTools version 2.36.0
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/dhs/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 65.1.4
    • Dart plugin version 213.7228

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • Darshan's iphone (mobile) • 21150b119064aecc249dfcfe05e259197461ce23 • ios
      • iOS 15.3.1 19D52
    • macOS (desktop)           • macos                                    •
      darwin-x64     • macOS 12.2.1 21D62 darwin-x64
    • Chrome (web)              • chrome                                   •
      web-javascript • Google Chrome 109.0.5414.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
      
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by luomo-pro** [comment]
Hello @vashworth , do you have time to take a look at this issue? Is it related to the changes you made in #94465 ?
In native iOS development, users don't need to focus on the keyboard to use the rotor for editing operations.

**Comment by lilangtop** [comment]
My app has received lots of feedback from blind users.
Blind users don't habitually focus the keyboard first, leading them to think they can't paste text.
I need to explain this to each blind user. Hope this issue can be fixed soon. Thanks!

**Comment by luomo-pro** [comment]
After continuous attempts, I seem to have found some solutions.

1. I confirmed that the modification in #94465 led to the emergence of this issue.
When I re-added UIAccessibilityTraitKeyboardKey, users don't need to focus on the keyboard to use operations like paste, select all, etc.
However, this causes the problem in #94465 to reappear.

2. I modified
flutter_engine/src/flutter/shell/platform/darwin/ios/framework/Source/TextInputSemanticsObject.mm
At the end of TextInputSemanticsObject, I added the following code:

```
- (void)cut:(id)sender {
    [[self textInputSurrogate] cut:sender];
}

- (void)copy:(id)sender {
    [[self textInputSurrogate] copy:sender];
}

- (void)paste:(id)sender {
    [[self textInputSurrogate] paste:sender];
}

- (void)select:(id)sender {
    [[self textInputSurrogate] select:sender];
}

- (void)selectAll:(id)sender {
    [[self textInputSurrogate] selectAll:sender];
}

- (void)delete:(id)sender {
    [[self textInputSurrogate] delete:sender];
}

	- (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
    return [[self textInputSurrogate] canPerformAction:action withSender:sender];
}
```

I found that it solved the problem!

3. However, the code above is not perfect.
In native iOS text fields, users don't need to activate the text field to perform operations like paste on the text field focused by VoiceOver.
But the code above doesn't work that way; it can only operate on already activated text fields.
Moreover, some options should not always be displayed. For example, "Paste" should only appear when there is text on the clipboard, and "Select All" should only be shown when there is text in the text field and it's not already fully selected, and so on.

I'm not sure if this modification has any negative impacts.
We need someone familiar with text fields to refine this, thank you!

**Comment by AlexV525** [comment]
cc @vashworth Could you help to identify fixes?

**Comment by chunhtai** [comment]
looks like @vashworth is out of office, cc @jmagman, do you have bandwidth to look into this?

**Comment by jmagman** [comment]
@hellohuanlin is going to take a look.

**Comment by hellohuanlin** [comment]
@luomo-pro @darshankawar i am looking into this. can you explain more in detail how I can reproduce this? 

What i did was that I single tap on a text field to select a text field, then triple tap it to "activate" it, which brings up the context menu. Then I swipe left or right to scroll through the menu items. 

I also tried rotor feature (2 finger rotate), and not sure how it's related to text context menu.

**Comment by luomo-pro** [comment]
@hellohuanlin

Please follow these steps:

1. Turn on VoiceOver.
2. Navigate to the text field and double-tap to activate it.
3. Ensure VoiceOver focus is on the text field.
Use the rotor gesture, navigate to "Edit", swipe down with one finger, you'll hear only "Scan Text" option.
4. Focus on the pop-up keyboard, e.g., the space key.
Use the rotor gesture again to navigate to "Edit", swipe down with one finger, you'll hear options like "Paste", "Select All", etc.

The issue is that users must focus on the keyboard to perform editing operations. If focused on the text field, editing operations are unavailable.

In contrast, iOS native text fields allow editing operations whether focused on the text field or keyboard.

To reproduce this issue, please use the latest Flutter master version.

**Comment by hellohuanlin** [comment]
I think the PR is good to go. Confirmed a few behaviors: 

> (1) Lack of VoiceOver feedback after operations.

This is existing behavior of the engine and not a regression by https://github.com/flutter/engine/pull/52333. But feel free to open a new ticket for that issue. 

> (2) Operations can only be performed on an activated text fields.

When I test it the native field still requires double tap to focus (iOS needs to know the cursor position in order to paste). Could you double check the behavior? 

(3) Some options should not always be displayed.

This is fixed in the PR. It was a bug in our FlutterTextInputPlugin (im surprised that this was not caught before). This also fixes the corresponding behavior for native edit menu (which is not enabled by default yet).

**Comment by luomo-pro** [comment]
@hellohuanlin
I apologize for the delay in responding; I've been away on business trips recently and just saw your comment.

Let me explain your second question:
(2) Operations can only be performed on an activated text field.
Suppose there are two text fields on an interface.
Double-tap to activate the first text field, and the keyboard pops up.
Then, use VoiceOver to focus on the second text field.
Note: just focus on it, don't double-tap.
Finally, use the rotor to perform a paste operation.

Flutter behavior: The text will be pasted into the first text field.
Native behavior: The text will be pasted into the second text field.

In other words, the native rotor operation targets the text field that VoiceOver is focused on.
However, Flutter rotor operation only targets the text field that has the keyboard activated.

**Comment by jmagman** [comment]
@luomo-pro the fix https://github.com/flutter/engine/pull/54351 is on the [master channel](https://github.com/flutter/flutter/blob/master/docs/releases/Flutter-build-release-channels.md#how-to-change-channels), if you want to try it out.
